### ISSUE TITLE
Number of adaptive Gaussian quadrature points in glmer()

### ISSUE THREAD

**Issue by nikola-sur** [lme4/lme4#668]
The current default for `nAGQ` is 1, resulting in a Laplace approximation in glmer(). Considering modern computation capabilities allowing for larger values of nAGQ and the possible inaccuracy of a Laplace approximation, would it be possible to increase this default?

**Comment by bbolker** [comment]
I think this is a bit of a tough call. Can I encourage you to post this question to `r-sig-mixed-models@r-project.org` for broader discussion? I can definitely see pros and cons

- safer/more conservative default
- not needed for cases with large effective cluster sizes, will slow down computations unnecessarily
- breaks backward compatibility

It's not available at present for models with non-scalar random effects terms (I've been meaning to do something for at least some special cases of this for a *long* time but it's not trivial ...)

**Comment by nikola-sur** [comment]
Yes, my main concern was backward compatibility. A discussion with some colleagues led to the thought that it might be a good idea to output a simple warning when cluster sizes are small if the default is to stay the same. Maybe just a note: `Warning: Small effective cluster sizes. Consider increasing nAGQ.` or something along those lines.

I think that some lay people using the package would not be aware that `nAGQ = 1` is the default and that it might impact the results. The warning might help without breaking backward compatibility and slowing down computations without need.

**Comment by bbolker** [comment]
Do you have concrete suggestions about computing effective cluster sizes?

- would depend on data type (`family`) (binomial: min(sum(successes), sum(failures)))
-  average cluster size across clusters?
- only apply when there's a single random effect term (I think this the only time when we can do AGQ anyway)? Or warn that cluster sizes are small with multiple random effect terms even though we don't have a solution? Warn if cluster sizes are small for *any* of the RE terms?
- would the estimation of effective cluster sizes be affected by having a vector rather than a scalar RE (i.e., random-slopes terms)?

**Comment by nikola-sur** [comment]
Thanks for the thoughtful questions. To be honest, I am a "casual user" myself, so I don't know if I am able to answer all of these appropriately.  

Maybe to only issue a warning when there is a single random effect term and the average cluster size is below some threshold? I'm not too sure at what cluster size an `nAGQ=1` approximation exactly starts to break down. If the problem requires more thought, I am fine with closing this issue and keeping the default as it is. I thought that I should just bring it up and see if others had thoughts.

**Comment by nikola-sur** [comment]
Thanks for the paper, Alex!

**Comment by bbolker** [comment]
This is indeed very nice.  A couple of quick questions:

- Is there any need to consider "effective cluster size", i.e. as far as I can tell m_i in this paper is the number of observations per cluster, whereas my previous **very** crude intuition had to do with some handwaving about the *effective* cluster size and convergence to Normality, i.e. a Bernoulli sample with low mean would need more quadrature points than a Bernoulli with a probability of 0.5, with similar arguments for a Poisson (k smaller when the mean is larger)
- I understand the reason for using `m = min(m_i)` to control error, but this strikes me as possibly pessimistic when there is a lot of variation in the cluster sizes (e.g. what if I have m = 200 for most clusters but m = 2 for a few?) ... in fact, there are a lot of practical situations where min(m_i) is 1 (i.e. we have some 'singleton' clusters) in which case everything goes to hell ...

**Comment by awstringer1** [comment]
Thanks for the questions and kind words!

- In practice, almost certainly yes. Your intuition of the convergence to Normality being what really matters matches mine as well. If the log-likelihoods are truly normal, the whole thing is exact for any k! The examples of Liu and Pierce (cited and discussed in Section 4 of our paper) are illuminating to me: they consider m = 1, and show one case where the integrand is almost exactly Gaussian by construction, and find that AGHQ performs well at any k. They then show one where it is not near to Gaussian, and find that AGHQ performs poorly for any k that they tried. This underscores your point, I think.
- This choice is definitely pessimistic, and this is for two reasons. 
    - First is that asymptotic arguments are often "worst-case"– the theory about AGHQ upon which our Theorem 2 is based quantifies the _worst_ possible rate at which such an integral of a "smooth" density can converge, and we matched this in spirit, trying to be as conservative as possible. I think we comment in Section 4 though that if min m_i = 1 then k = infinity, because of this, and this underscores the potential limitations of relying solely on asymptotic recommendations. 
    - Second is that the error in the total likelihood approximation is multiplicative: you have M integrals being approximated then multiplied, so I would expect the worst one to have a negative effect that does not improve (gets worse?) as M gets bigger, and for this to arise in "typical" applications rather than just worst-case ones. In the case you have some m_i = 1, I think the whole thing is then limited by "how Gaussian" those clusters' joint likelihoods already are– the sample just doesn't help you (asymptotically).
